down-the-stack/dk_template/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+}
+
+/// Exits the application when the program is executed through the `probe-run` Cargo runner
+pub fn exit() -> ! {
+    unsafe {
+        // turn off the USB D+ pull-up before pausing the device with a breakpoint
+        // this disconnects the nRF device from the USB host so the USB host won't attempt further
+        // USB communication (and see an unresponsive device). probe-run will also reset the nRF's
+        // USBD peripheral when it sees the device in a halted state which has the same effect as
+        // this line but that can take a while and the USB host may issue a power cycle of the USB
+        // port / hub / root in the meantime, which can bring down the probe and break probe-run
+        const USBD_USBPULLUP: *mut u32 = 0x4002_7504 as *mut u32;
+        USBD_USBPULLUP.write_volatile(0)
+    }
+    defmt::println!("`dk::exit()` called; exiting ...");
+    // force any pending memory operation to complete before the BKPT instruction that follows
+    atomic::compiler_fence(Ordering::SeqCst);
+    loop {
+        asm::bkpt()
+    }
+}
+
+/// Returns the time elapsed since the call to the `dk::init` function
+///
+/// The clock that is read to compute this value has a resolution of 30 microseconds.
+///
+/// Calling this function before calling `dk::init` will return a value of `0` nanoseconds.
+pub fn uptime() -> Duration {
+    // here we are going to perform a 64-bit read of the number of ticks elapsed
+    //
+    // a 64-bit load operation cannot performed in a single instruction so the operation can be
+    // preempted by the RTC0 interrupt handler (which increases the OVERFLOWS counter)
+    //
+    // the loop below will load both the lower and upper parts of the 64-bit value while preventing
+    // the issue of mixing a low value with an "old" high value -- note that, due to interrupts, an
+    // arbitrary amount of time may elapse between the `hi1` load and the `low` load
+    let overflows = &OVERFLOWS as *const AtomicU32 as *const u32;
+    let ticks = loop {
+        unsafe {
+            // NOTE volatile is used to order these load operations among themselves
+            let hi1 = overflows.read_volatile();
+            let low = core::mem::transmute::<_, RTC0>(())
+                .counter
+                .read()
+                .counter()
+                .bits();
+            let hi2 = overflows.read_volatile();
+
+            if hi1 == hi2 {
+                break u64::from(low) | (u64::from(hi1) << 24);
+            }
+        }
+    };
+
+    // 2**15 ticks = 1 second
+    let freq = 1 << 15;
+    let secs = ticks / freq;
+    // subsec ticks
+    let ticks = (ticks % freq) as u32;
+    // one tick is equal to `1e9 / 32768` nanos
+    // the fraction can be reduced to `1953125 / 64`
+    // which can be further decomposed as `78125 * (5 / 4) * (5 / 4) * (1 / 4)`.
+    // Doing the operation this way we can stick to 32-bit arithmetic without overflowing the value
+    // at any stage
```

Comment:
I would just use a u64 or a u128 and let the compiler figure it out. It'd take me at least 10 minutes to satisfy myself this was correct and use most of a sheet of A4 paper doing it.